cuda/matrix/dense_kernels.cu

```diff
@@ -346,20 +349,199 @@ void compute_norm2(std::shared_ptr<const CudaExecutor> exec,
 
 GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_DENSE_COMPUTE_NORM2_KERNEL);
 
+namespace kernel {
+
+__global__ __launch_bounds__(cuda_config::warp_size) void start_nnz_sum(
+    size_type num_rows, size_type work_size,
+    size_type *__restrict__ nnz_per_row, size_type *__restrict__ add_values)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+    if (tidx < num_rows) {
+        auto start_row = tidx * work_size;
+        auto nnz_up_to_here = nnz_per_row[start_row];
+        for (auto i = start_row + 1; i < min(start_row + work_size, num_rows);
+             i++) {
+            nnz_up_to_here += nnz_per_row[i];
+            nnz_per_row[i] = nnz_up_to_here;
+        }
+        add_values[tidx] = nnz_up_to_here;
+    }
+}
+
+__global__ __launch_bounds__(default_block_size) void finalize_nnz_sum(
+    size_type num_rows, size_type *__restrict__ nnz_per_row,
+    size_type *__restrict__ add_values)
+{
+    const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
+
+    if (tidx + blockDim.x < num_rows) {
+        size_type add_value = 0;
+        for (auto i = 0; i <= blockIdx.x; i++) {
+            add_value += add_values[i];
+        }
+        nnz_per_row[tidx + blockDim.x] += add_value;
+    }
+}
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void fill_in_coo(
+    size_type num_rows, size_type num_cols, size_type stride,
+    const size_type *__restrict__ row_ptrs,
+    const ValueType *__restrict__ source, IndexType *__restrict__ row_idxs,
+    IndexType *__restrict__ col_idxs, ValueType *__restrict__ values)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+    if (tidx < num_rows) {
+        size_type write_to = (tidx == 0) ? 0 : row_ptrs[tidx - 1];
+
+        for (auto i = 0; i < num_cols; i++) {
+            if (source[stride * tidx + i] != zero<ValueType>()) {
+                values[write_to] = source[stride * tidx + i];
+                col_idxs[write_to] = i;
+                row_idxs[write_to] = tidx;
+                write_to++;
+            }
+        }
+    }
```

Comment:
I think if we calculate col_ptrs instead of row_ptrs, we end up with sorted col_idxs instead of sorted row_idxs, right? So to get a column major coo matrix we have to sort everything which would be pretty expensive.

Reply:
It is fine as it is right now, I had an idea which would cost us more memory and I don't think it is worth it.